fix (controls): positioning error during camera movements

gchoqueux opened the pull request:
There was positioning error during camera movements with `globeControls`

**First commit**
In `setCameraTargetPosition` without animation, the final position was wrong because the transformation was simply not applied.
Can be tested through the function `setCameraTargetGeoPosition`

**second commit**
Other positioning error come from with double click.
Because it's in conflict with the simple click that calls the `setCameraTargetPosition` too.
For double click event:
  * with animation enabled problem is fixed
  * with animation disabled is forced to enable

gchoqueux committed:
fix (controls): in GlobeControls.setCameraTargetPosition the final position isn't the desired position

without animation, the final position was wrong because the transformation was simply not applied

autra commented:
> In PR #545 

#545 is an issue, you meant #535 maybe ?

gchoqueux commented:
yes, thanks #535

autra reviewed (changes requested):
Commit message should preferably explain why the change (and not what, most of the time).

So what is the actual bug? Can you give me reproduction steps, expected and actual results please? I tried to double-click everywhere with animation disabled (this leads to this new code being executed), but couldn't see the difference before and after your patch: this double-click is still completely broken in both cases...

gchoqueux commented:
@autra 
Thanks, for the bug with double-click.
I updated PR's message and the commit

gchoqueux commented:
@autra PR is updated

nosy-b commented:
Weird behaviour when leaving the focus (opening a terminal) then going back to itowns, the click is no more listened, only with CTRL - CLICK I can see that the app is not freezed.
(I specified enableAnimation to false, it still animates translation also...)

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1114,7 +1114,7 @@ function GlobeControls(view, target, radius, options = {}) {
             if (point) {
                 animatedScale = 0.6;
-                this.setCameraTargetPosition(point, this.isAnimationEnabled());
+                this.setCameraTargetPosition(point, true);
```
This was for testing right? If so should be remoed

gchoqueux replied:
I added comment, there's conflict simple/double click without animation. I'll fix it in the refacto.

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1619,13 +1619,17 @@ GlobeControls.prototype.setCameraTargetPosition = function setCameraTargetPositi
     } else {
         ctrl.progress = 1.0;
         quatGlobe.setFromUnitVectors(vFrom, vTo);
+        this.updateCameraTransformation(this.states.MOVE_GLOBE, false);
```
MOVE_GLOBE is the default state?

gchoqueux replied:
Default state is `NONE`, in `setCameraTargetPosition uses `MOVE_GLOBE` to move target's position`

autra replied:
Ok, not very intuitive (we might want another name for the refacto), but fair enough.

autra reviewed (changes requested):
One leftover+a question  and a check on  @nosy-b's comments.

After this is done, I'll do one last test and I guess we can merge that. I'm not sure to really understand what's going on here though...

autra commented:
@nosy-b just tested on master and reproduced the 2 bugs you mentioned. They are not caused by this PR.

gchoqueux commented:
@nosy-b fix focus problem

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1422,6 +1425,12 @@ function GlobeControls(view, target, radius, options = {}) {
     window.addEventListener('keyup', onKeyUp.bind(this), false);
 
+    // Reset key/mouse when window loose focus
+    window.onblur = () => {
```
`this.domElement.addEventListener('blur',...)`: this will fix it it another element takes focus somehow too. Also, maybe the webpage in which itowns is integrated uses also this blur event. Here you're overriding whatever might already be in `window.onblur`.

Also, now that the review process has begun, I'd prefer you open new PR for new fixes that come on the way.

gchoqueux replied:
I change in `window.addEventListener('blur'`...
`this.domElement.addEventListener` doesn't work

autra replied:
Ok works for me. (it won't fix it if another element takes focus in the page, but that's not the goal of this PR)

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1114,7 +1114,10 @@ function GlobeControls(view, target, radius, options = {}) {
+                // FIXME: single click and double click are in conflict with no animation
+                // with this.setCameraTargetPosition
+                // force animation at true to tempory fix the problem
+                this.setCameraTargetPosition(point, true);
```
If it works with animation, but not without, it looks like a race condition. Anyhow I'm skeptical of this fix. Adding an animation when animations are disabled to fix a bug is... not exactly the most straightforward thing I can think of. I'm sure there's a way to make it work with animation disabled. 

What is the conflict exactly?  Can you describe the bug in more details?

gchoqueux replied:
I don't want to spend more time on this version,
I will correct that in the refacto

autra replied:
Then if you don't want to spend more time here, let's deactivate the double-click action when enableAnimation is false. We can't really merge code that lies! (doing an animation when enableAnimation is false)

If this does not suit you, then I'll try to find a solution when I have time, but you'll have to wait for me. If you want to speed up this process, you can answer my questions 

> What is the conflict exactly? Can you describe the bug in more details?

If you don't want to wait, you'll have to propose an alternative solution :-)

autra replied:
Anyway, maybe it's better to put the problematic commit into its own PR. This way, I'll merge this one now.

gchoqueux replied:
I disable without animation and I add message warning

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1114,7 +1120,10 @@ function GlobeControls(view, target, radius, options = {}) {
-                this.setCameraTargetPosition(point, this.isAnimationEnabled());
+                // FIXME: single click and double click are in conflict with no animation
+                // with this.setCameraTargetPosition
+                // force animation at true to tempory fix the problem
```
This comment is now outdated. Let's remove it.

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1105,6 +1105,12 @@ function GlobeControls(view, target, radius, options = {}) {
 
+        if (!this.isAnimationEnabled()) {
+             // eslint-disable-next-line no-console
+            console.warn('double click without animation is disabled, waiting fix in new factorization');
```
`waiting fix in future refactoring`?

autra reviewed (changes requested):
The new diff is on the wrong commit (is in last commit, should be on second). Also, second commit message needs to be updated. After this review, it'll be ready to merge.

gchoqueux commented:
@autra done

gchoqueux commented:
@autra we merge it?

autra commented:
@gchoqueux I think you forgot to adress

> The new diff is on the wrong commit (is in last commit, should be on second). Also, second commit message needs to be updated. After this review, it'll be ready to merge.

So I took the liberty to amend your commits. Please double-check if it's ok for you, then we can merge.

autra commented:
(I have a backup of the old commits if needed)

gchoqueux committed:
fix (controls): conflict between simple click and double click

double click and simple click events call setCameraTargetPosition.
There was a conflict, which causes positioning errors.
For double click event:
  * with animation enabled problem is fixed
  * if animation disabled, the double click event is disabled too (waiting fix in future refactoring)

gchoqueux committed:
fix (controls): reset globeControls when window's focus is losed

autra commented:
Ok the diff are alright, but the commit message needs some amending. You can take my proposals or not, but at least on second commit message:

>   * with animation disabled is forced to enable

should be changed, as it's not the case anymore. Thanks.

autra reviewed (approved):
Here we go!
